hilti/src/rt/library.cc

```diff
@@ -32,13 +69,28 @@ hilti::rt::Library::~Library() {
         hilti::rt::warning(fmt("could not remove library %s from store: %s", ec.message()));
 }
 
-hilti::rt::Result<hilti::rt::Nothing> hilti::rt::Library::open() const {
+hilti::rt::Result<hilti::rt::library::Version> hilti::rt::Library::open() const {
     constexpr auto mode = RTLD_LAZY | RTLD_GLOBAL;
 
-    if ( ! ::dlopen(_path.c_str(), mode) )
+    void* hlto = ::dlopen(_path.c_str(), mode);
+    if ( ! hlto )
         return result::Error(fmt("failed to load library %s: %s", _path, dlerror()));
 
-    return hilti::rt::Nothing();
+    auto version_string = reinterpret_cast<const char**>(::dlsym(hlto, "__hlto_library_version"));
+    if ( ! version_string )
+        return result::Error("no version information");
+
+    library::Version version;
+    if ( auto rc = version.fromJSON(*version_string); ! rc )
+        return result::Error(fmt("broken version information (%s)", rc.error()));
+
+    // Check version. We only warn for now, don't abort.
+    if ( version.magic != "v1" )
+        result::Error(fmt("unknown HLTO version '%s'", version.magic));
+
+    version.path = _orig_path;
+    version.checkCompatibility();
+    return version;
```

Comment:
We could move here to avoid the copy in the conversion to the return type.
```suggestion

    return std::move(version);
```